File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/basicfunctionality/MotorControl.java
package org.firstinspires.ftc.teamcode.hardware.basicfunctionality;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;

import org.firstinspires.ftc.teamcode.hardware.RobotHardware;

public class MotorControl {

  private RobotHardware rh;
  private DcMotorEx motor;
  private String name;

  private int min = 0;
  private int max = 0;

  private int[] positions = null;

  // These separate increasing and decreasing speeds for cases such as
  // going slower down so that gravity doesn't make a lift system damage the robot
  private double increasingSpeed = 1.0;
  private double decreasingSpeed = 1.0;

  // This initializes the MotorControl regardless of constructor type
  private void construct(RobotHardware rh, String motorName, DcMotorEx.Direction direction) {
    this.rh = rh;
    motor = rh.op.hardwareMap.get(DcMotorEx.class, motorName);
    name = motorName;

    motor.setTargetPosition(0);

    motor.setMode(DcMotorEx.RunMode.RUN_USING_ENCODER);
    motor.setDirection(direction);
  }

  // Default ZeroPowerBehavior is BRAKE
  public MotorControl(RobotHardware rh, String motorName, DcMotorEx.Direction direction) {
    construct(rh, motorName, direction);
    motor.setZeroPowerBehavior(DcMotorEx.ZeroPowerBehavior.BRAKE);
  }

  public MotorControl(RobotHardware rh, String motorName, DcMotorEx.Direction direction, DcMotorEx.ZeroPowerBehavior zeroPowerBehavior) {
    construct(rh, motorName, direction);
    motor.setZeroPowerBehavior(zeroPowerBehavior);
  }

  public void move() { motor.setPower(powerCurve()); }

  public boolean isMoving() {
    if (dumbMode) return 0.0 != motor.getPower();
    else return 0.0 != (motor.getCurrentPosition() - motor.getTargetPosition());
  }

  public void setRange(int min, int max) {
    this.min = min;
    this.max = max;
  }

  public void setPositions(int[] positions) { this.positions = positions; }

  public void goToPresetPosition(int index) { motor.setTargetPosition(positions[index]); }

  public void clampTargetInRange() {
    int target = Math.min(max, motor.getTargetPosition());
    target = Math.max(min, target);
    motor.setTargetPosition(target);
  }

  // --------------------------------------------------------------------------------------------

  private boolean dumbMode = true;

  public boolean isDumbMode() { return dumbMode; }

  public void disableEncoder() {
    dumbMode = true;
    motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
  }

  public void enableEncoder() {
    dumbMode = false;
    motor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    motor.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
    motor.setTargetPosition(motor.getCurrentPosition());
  }

  public void moveDumb(double power) {
    motor.setPower(power);
  }

  // --------------------------------------------------------------------------------------------

  public void goToPosition(int targetPosition) { motor.setTargetPosition(targetPosition); }

  public int getCurrentPosition() { return motor.getCurrentPosition(); }

  public int getTargetPosition() { return motor.getTargetPosition(); }

  // --------------------------------------------------------------------------------------------

  public void setSpeedControls(double speed) {
    this.increasingSpeed = speed;
    this.decreasingSpeed = speed;
  }

  public void setSpeedControls(double increasingSpeed, double decreasingSpeed) {
    this.increasingSpeed = increasingSpeed;
    this.decreasingSpeed = decreasingSpeed;
  }

  public void setPowerControls(int easingDistance, int brakingDistance, double stretch) {
    this.easingDistance = easingDistance;
    this.brakingDistance = brakingDistance;
    this.stretch = stretch;
  }

  // variables only for the power curve method
  private int easingDistance = 1000;      // Max speed is used above this distance
  private int brakingDistance = 30;       // Brake below this distance from target
  private double stretch = 100;           // For a base of 2 this is the distance at which 50% speed will be used
  // Above this distance it will approach 100% speed
  // Below this distance it will approach 0% speed with a deadzone near 0 distance

  public double powerCurve() {
    if (dumbMode) return 0.0;

    // find signed distance to target
    int distance = motor.getTargetPosition() - motor.getCurrentPosition();

    // set the correct speed for the direction of motion
    double speed = increasingSpeed;
    if (distance < 0) {
      speed = decreasingSpeed;
    }

    // make the distance positive
    distance = Math.abs(distance);

    // returns 0.0 when within braking distance
    if (distance <= brakingDistance) {
      return 0.0;
    }

    // returns speed when above easing distance
    if (distance >= easingDistance) {
      return speed;
    }

    // when inside easing distance, use a decreasing speed for closer distances
    // the exponential base (2) shouldn't be changed
    double power = Math.pow(2, -stretch / distance);
    return power * speed;

    // y = 2^( -s / x ) * v
    // where y is power, x is distance to target (+), s is the stretch factor, and v is the speed
  }

  public void telemetry() {
    rh.op.telemetry.addLine();

    rh.op.telemetry.addLine(String.format("Motor %s:", name));

    rh.op.telemetry.addLine(String.format("    MIN is %d, MAX is %d", min, max));
    rh.op.telemetry.addLine(String.format("    Positions: CURRENT %d, TARGET %d", motor.getCurrentPosition(), motor.getTargetPosition()));
    rh.op.telemetry.addLine(String.format("    %s moving%s", isMoving() ? "IS" : "NOT", isMoving() ? String.format(" at speed %d", powerCurve()) : "")); // Gross.
  }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/basicfunctionality/MotorControl.kt
package org.firstinspires.ftc.teamcode.hardware.basicfunctionality

import com.qualcomm.robotcore.hardware.DcMotor.RunMode
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction
import org.firstinspires.ftc.teamcode.hardware.RobotHardware
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

/**
 * A motor, generally one designed to operate with encoders. For example, a lift.
 *
 * This class provides functionalities beyond the basic `DcMotorEx` interface,
 * including:
 * - Setting operational range (min/max encoder positions).
 * - Defining preset positions for quick navigation.
 * - Clamping target positions within the defined range.
 * - Toggling between "dumb" (direct power control) and encoder-based movement.
 * - Implementing a power curve for smoother acceleration and deceleration.
 * - Separate speed controls for increasing and decreasing movements (e.g., for lifts).
 * - Telemetry output for debugging and monitoring.
 *
 * @property rh An instance of [RobotHardware] providing access to the op mode and hardware map.
 * @property name The name of the motor as set in the hardware configuration on the driver hub.
 * @property direction The [Direction] of the motor (FORWARD or REVERSE).
 * @property zeroPowerBehavior (Optional) The [ZeroPowerBehavior] of the motor (BRAKE or FLOAT) when power is zero.
 *                             Defaults to BRAKE.
 */
class MotorControl(val rh: RobotHardware, val name: String, val direction: Direction, val zeroPowerBehavior: ZeroPowerBehavior) {
  private var motor: DcMotorEx = rh.op.hardwareMap.get(DcMotorEx::class.java, name)

  private var min = 0
  private var max = 0

  private var positions: IntArray = intArrayOf()

  // These separate increasing and decreasing speeds for cases such as
  // going slower down so that gravity doesn't make a lift system damage the robot
  private var increasingSpeed = 1.0
  private var decreasingSpeed = 1.0

  init {
    motor.targetPosition = 0

    motor.mode = RunMode.RUN_USING_ENCODER
    motor.direction = direction
    motor.zeroPowerBehavior = zeroPowerBehavior
  }

  // Default ZeroPowerBehavior is BRAKE
  constructor(rh: RobotHardware, name: String, direction: Direction) : this(rh, name, direction, ZeroPowerBehavior.BRAKE)

  fun move() {
    motor.power = powerCurve()
  }

  val isMoving: Boolean
    get() {
      return if (this.isDumbMode) 0.0 != motor.power
      else 0.0 != (motor.currentPosition - motor.targetPosition).toDouble()
    }

  fun setRange(min: Int, max: Int) {
    this.min = min
    this.max = max
  }

  fun setPositions(positions: IntArray) {
    this.positions = positions
  }

  fun goToPresetPosition(index: Int) {
    motor.targetPosition = positions[index]
  }

  fun clampTargetInRange() {
    var target = min(max, motor.targetPosition)
    target = max(min, target)
    motor.targetPosition = target
  }

  // --------------------------------------------------------------------------------------------
  var isDumbMode: Boolean = true
    private set

  fun disableEncoder() {
    this.isDumbMode = true
    motor.mode = RunMode.RUN_WITHOUT_ENCODER
  }

  fun enableEncoder() {
    this.isDumbMode = false
    motor.mode = RunMode.STOP_AND_RESET_ENCODER
    motor.mode = RunMode.RUN_USING_ENCODER
    motor.targetPosition = motor.currentPosition
  }

  fun moveDumb(power: Double) {
    motor.power = power
  }

  // --------------------------------------------------------------------------------------------
  fun goToPosition(targetPosition: Int) {
    motor.targetPosition = targetPosition
  }

  val currentPosition: Int
    get() = motor.currentPosition

  val targetPosition: Int
    get() = motor.targetPosition

  // --------------------------------------------------------------------------------------------
  fun setSpeedControls(speed: Double) {
    this.increasingSpeed = speed
    this.decreasingSpeed = speed
  }

  fun setSpeedControls(increasingSpeed: Double, decreasingSpeed: Double) {
    this.increasingSpeed = increasingSpeed
    this.decreasingSpeed = decreasingSpeed
  }

  fun setPowerControls(easingDistance: Int, brakingDistance: Int, stretch: Double) {
    this.easingDistance = easingDistance
    this.brakingDistance = brakingDistance
    this.stretch = stretch
  }

  // Variables only for the power curve method
  private var easingDistance = 1000 // Max speed is used above this distance
  private var brakingDistance = 30 // Brake below this distance from target
  private var stretch = 100.0 // For a base of 2 this is the distance at which 50% speed will be used

  // Above this distance it will approach 100% speed
  // Below this distance it will approach 0% speed with a deadzone near 0 distance
  fun powerCurve(): Double {
    if (this.isDumbMode) return 0.0

    // Find signed distance to target
    var distance = motor.targetPosition - motor.currentPosition

    // Set the correct speed for the direction of motion
    var speed = increasingSpeed
    if (distance < 0) {
      speed = decreasingSpeed
    }

    // Make the distance positive
    distance = abs(distance)

    // Returns 0.0 when within braking distance
    if (distance <= brakingDistance) {
      return 0.0
    }

    // Returns speed when above easing distance
    if (distance >= easingDistance) {
      return speed
    }

    // When inside easing distance, use a decreasing speed for closer distances
    // The exponential base (2) shouldn't be changed
    val power = 2.0.pow(-stretch / distance)
    return power * speed

    // y = 2^( -s / x ) * v
    // Where y is power, x is distance to target (+), s is the stretch factor, and v is the speed
  }

  fun telemetry() {
    rh.op.telemetry.addLine("\nmotor '$name'")
    rh.op.telemetry.addLine("    MIN is $min, MAX is $max")
    rh.op.telemetry.addLine("    positions: CURRENT ${motor.currentPosition}, TARGET ${motor.targetPosition}")
    rh.op.telemetry.addLine("    ${if (this.isMoving) "IS" else "NOT"} moving${if (this.isMoving) " at speed ${powerCurve()}" else ""}")
  }
}
